### PR TITLE
workload registrar: federation support

### PR DESCRIPTION
**Pull request by mazamats**
**Pull Request check list**

- [x] Commit conforms to CONTRIBUTING.md?
- [x] Proper tests/regressions included?
- [x] Documentation updated?

**Affected functionality**
This adds the ability to create SPIRE entries that federate with other trust domains.

**Description of change**
Look for the pod annotation `spiffe.io/federatesWith`, comma separate the value, use that to create the entry
 .
**Which issue this PR fixes**
N/A

### CODE REVIEW

**Review by azdagron — COMMENTED** [review]
Thanks for this @mazamats. I apologize for the latency on the review.

Currently, the added configurable seems to control two different behaviors: 1) the opting into the adding of FederatesWith by specifying the configurable, and 2) naming the annotation that we look at to gather the FederatesWith values.

I wonder instead if we can just hard code the annotation that we look at (i.e., `spiffe.io/federatesWith`) and provide the opt-in/out functionality with a simple boolean configurable, that defaults to a sane value (probably default to on?).

I'd advocate for no configurable at all, but maybe there are situations where people can influence pod annotations that should not be influencing this behavior and an operator may wish to disable this.

Reducing the configuration landscape is something we're trying to get better at.

**Comment by mazamats** [comment]
> Reducing the configuration landscape is something we're trying to get better at.

Glad to hear that, I implemented that way initially but added the config when I saw the other parameters to modify annotations.

I'd advocate for no configurable as well, I can't see a scenario where a cluster operator would have that annotation set without actually needed federation.

File: support/k8s/k8s-workload-registrar/mode-crd/api/spiffeid/v1beta1/spiffeid_types.go [hunk 1]
```diff
@@ -47,10 +47,11 @@ type Selector struct {
...
 type SpiffeIDSpec struct {
-	ParentId string   `json:"parentId"`
-	SpiffeId string   `json:"spiffeId"`
-	Selector Selector `json:"selector"`
-	DnsNames []string `json:"dnsNames,omitempty"`
+	ParentId      string   `json:"parentId"`
+	SpiffeId      string   `json:"spiffeId"`
+	Selector      Selector `json:"selector"`
+	DnsNames      []string `json:"dnsNames,omitempty"`
+	FederatesWith []string `json:"federatesWith,omitempty"`
```
Comment (faisal-memon) [inline support/k8s/k8s-workload-registrar/mode-crd/api/spiffeid/v1beta1/spiffeid_types.go:54]: I think this will require updating the api version. we are planning to do this anyways, so perhaps we can lump this in so we don't have to update too many times?

File: support/k8s/k8s-workload-registrar/controller.go [hunk 2]
```diff
@@ -153,7 +152,7 @@ func (c *Controller) createPodEntry(ctx context.Context, pod *corev1.Pod) error
 }
 
 func (c *Controller) getFederationDomains(pod *corev1.Pod) []string {
-	if val, ok := pod.GetAnnotations()[c.c.FederationAnnotation]; ok {
+	if val, ok := pod.GetAnnotations()["spiffe.federatesWith"]; ok {
```
Comment (azdagron) [inline support/k8s/k8s-workload-registrar/controller.go:155]: The previous default in the PR was `spiffe.io/federatesWith`? Why the change to `spiffe.federatesWith`?

Also, this value should be a constant somewhere.

File: support/k8s/k8s-workload-registrar/controller_test.go [hunk 3]
```diff
@@ -387,6 +421,72 @@ func TestControllerAnnotationBasedRegistration(t *testing.T) {
...
+		Namespace: "NAMESPACE",
+		Name:      "PODNAME",
+		Operation: "CREATE",
+		Object: runtime.RawExtension{
+			Raw: []byte(fakePodWithFederation),
+		},
+	})
+
+	// Assert that the registration entry for the pod was created
+	requireEntriesEqual(t, []*types.Entry{
```
Comment (azdagron) [inline support/k8s/k8s-workload-registrar/controller_test.go:443]: Something here is causing infinite recursion and causing the test to run out of memory...

File: support/k8s/k8s-workload-registrar/mode-crd/api/spiffeid/v1beta1/spiffeid_types.go (hunk 1 above)
Comment (mazamats) [reply support/k8s/k8s-workload-registrar/mode-crd/api/spiffeid/v1beta1/spiffeid_types.go:54]: I'm not planning to use the CRD mode, is it ok if we push this through without updating the API version and it gets picked up later?

File: support/k8s/k8s-workload-registrar/controller.go (hunk 2 above)
Comment (mazamats) [reply support/k8s/k8s-workload-registrar/controller.go]: I moved it all to its own package

File: support/k8s/k8s-workload-registrar/mode-crd/api/spiffeid/v1beta1/spiffeid_types.go (hunk 1 above)
Comment (faisal-memon) [reply support/k8s/k8s-workload-registrar/mode-crd/api/spiffeid/v1beta1/spiffeid_types.go:54]: ok with me

**Review by azdagron — APPROVED** [review]
Thanks for this, @mazamats !
